### PR TITLE
Manual sync of files pressing p

### LINKED ISSUE #5634: manually trigger file sync in `odo dev`
## User Story
- As an odo user running `odo dev --no-watch`
- I want to be able to trigger push and restart on demand
- So that I can trigger push only after I made multiple edits, because I don't need to see every single small change and on a complex application it can be expensive to trigger rebuild

## Acceptance Criteria

1. This issue
- [ ] when `odo dev` is running user should be able to trigger push by pressing a key (p).
- [ ] `odo dev` output should inform user about this option
- [ ] triggering push on-demand should be possible regardless if `odo dev` is running with or without `--no-watch`

2. How to trigger the file sync
 https://github.com/redhat-developer/odo/issues/5986

/kind user-story

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

This PR configures the terminal to work in Character mode, instead of the default Line mode. Read() will return as soon as a character is sent to the input stream, instead of waiting for a newline character.

code refactoring:
- moved getFullSourcesWatcher and related functions to a specific file file_watcher.go
- moved long list of watchers variables into receiver instead of passing as parameters

**Which issue(s) this PR fixes:**

Fixes #5634

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

- `odo dev` will fail to configure terminal on Windows/cygwin terminal.

  You can use `$ winpty odo dev` to fix the problem

  https://bryantson.medium.com/troubleshooting-make-raw-terminal-failed-the-handle-is-invalid-57b7e1d9b12b

- Windows terminal seetings
  https://docs.microsoft.com/en-us/windows/console/setconsolemode

- Unix terminal settings:
  https://man7.org/linux/man-pages/man3/termios.3.html

## Refactoring integration tests

Some tests were calling devSession.Kill() (or Stop()) twice, which makes ginkgo stop abrupty during the second call on Windows.

In the following example, the Kill in AfterEach is necessary for the cleanup after test1, but is problematic for test2, as Stop has been called before.

As the AfterEach/BeforeEach are evaluated at the beginning, it is not possible to rely on a field in devSession to record if the method has already been called.

```
When("odo dev is executed", func() {

			var devSession helper.DevSession

			BeforeEach(func() {
				var err error
				devSession, _, _, _, err = helper.StartDevMode(nil)
				Expect(err).ToNot(HaveOccurred())
			})

			AfterEach(func() {
				devSession.Kill()
				devSession.WaitEnd()
			})

			It("test 1", func() { ... })

			When("odo dev is stopped", func() {
				BeforeEach(func() {
					devSession.Stop()
					devSession.WaitEnd()
				})

				It("test 2", func() { ... })
			})
		})
```

These tests have been split in two trees, separating test1 and test2, to be able to call Kill/Stop only once.

### CODE REVIEW

**Comment by feloy** [comment]
Blocked by #6091

**Comment by feloy** [comment]
Thanks @rm3l for your review

File: pkg/watch/watch.go
Comment (rm3l) [inline pkg/watch/watch.go:40]: One last thing I was thinking of: maybe we could also say that pressing 'p' might also restart/reload the running application.
This could be mentioned here and/or in the documentation for `odo dev`.
Other than sync'ing source files, I didn't see anything related to the running application mentioned (unless that's what we mean by "sync").

File: docs/website/docs/command-reference/dev.md
Comment (rm3l) [inline docs/website/docs/command-reference/dev.md:61]: `HotReloadCapable` is a field at the command level, not the component.

**Review by rm3l — APPROVED** [review]
Thanks for the changes.
LGTM. I'll let other reviewers take another look.

**Comment by dharmit** [comment]
Unrelated to the code changes in this PR. I noticed that the `watch.go` file has many timers now. 

```go
sourcesTimer := time.NewTimer(time.Millisecond)
<-sourcesTimer.C

devfileTimer := time.NewTimer(time.Millisecond)
<-devfileTimer.C

deployTimer := time.NewTimer(time.Millisecond)
<-deployTimer.C

retryTimer := time.NewTimer(time.Millisecond)
<-retryTimer.C
```
IMO, it would help if we can add a comment that points which watcher it corresponds to. I'm OK if we want to open a new issue and mark it as good first issue (Hacktoberfest is just around the corner.)

File: pkg/watch/key_watcher_windows.go
Comment (dharmit) [inline pkg/watch/key_watcher_windows.go:7]: What is the `makeRaw` that you refer to here? I don't find it anywhere except this comment:
```shell
$  grep -r makeRaw pkg/      
pkg/watch/key_watcher_windows.go:// makeRaw is forked from golang.org/x/term

```

File: pkg/watch/term_unix_other.go
Comment (dharmit) [inline pkg/watch/term_unix_other.go:3]: Is it OK to have this when the odo code is licensed as Apache Licence 2.0?

File: pkg/watch/term_unix_other.go
Comment (dharmit) [inline pkg/watch/term_unix_other.go:13]: What are these constants for exactly? 
What's the significance of this file and the file `term_unix_bsd.go`?

File: pkg/watch/key_watcher_unix.go
Comment (dharmit) [inline pkg/watch/key_watcher_unix.go:21]: I find this to be magic. Can you explain in code comment as to what this is happening here?

**Comment by feloy** [comment]
> Unrelated to the code changes in this PR. I noticed that the `watch.go` file has many timers now.
> 
> ```go
> sourcesTimer := time.NewTimer(time.Millisecond)
> <-sourcesTimer.C
> 
> devfileTimer := time.NewTimer(time.Millisecond)
> <-devfileTimer.C
> 
> deployTimer := time.NewTimer(time.Millisecond)
> <-deployTimer.C
> 
> retryTimer := time.NewTimer(time.Millisecond)
> <-retryTimer.C
> ```
> 
> IMO, it would help if we can add a comment that points which watcher it corresponds to. I'm OK if we want to open a new issue and mark it as good first issue (Hacktoberfest is just around the corner.)

I've added some comments

File: pkg/watch/file_watcher.go
Comment (dharmit) [inline pkg/watch/file_watcher.go:1]: @feloy, there's a bunch of things being added/moved under the `pkg/watch` directory. I humbly request you to please add comments to the code. Especially functions, even if they are internal. It feels like magic right now, and could be really difficult for others to comprehend what's going on.

**Review by dharmit — CHANGES_REQUESTED** [review]

File: pkg/watch/key_watcher_windows.go
Comment (feloy) [reply pkg/watch/key_watcher_windows.go]: `enableCharInput` just below. Sorry I renamed the function name, but didn't update the comment

File: pkg/watch/term_unix_other.go
Comment (feloy) [reply pkg/watch/term_unix_other.go:3]: It is the license for the Go library and the code coming from `golang.org/x/*` repos.
I can imagine it is ok

This code specifically comes from the `golang.org/x/term` repository (https://cs.opensource.google/go/x/term/+/a9ba230a:term_unix_other.go)

File: pkg/watch/key_watcher_unix.go
Comment (feloy) [reply pkg/watch/key_watcher_unix.go:24]: I've put links on the PRs for you to understand what it is about:

https://man7.org/linux/man-pages/man3/termios.3.html

This is not magic, but terminal settings.

File: pkg/watch/file_watcher.go
Comment (feloy) [reply pkg/watch/file_watcher.go:1]: I've added some more comments to the functions. But note that I added into the PR comment links to Unix and Windows terminal settings reference, that will help you understand what it is about.

File: pkg/watch/term_unix_other.go
Comment (feloy) [reply pkg/watch/term_unix_other.go:13]: They are the constants for Terminal I/Os on Unix. Unfortunately, constants values are different on BSD and on other Unixes, thus the 2 different files with different constants.

File: pkg/watch/file_watcher.go
Comment (dharmit) [reply pkg/watch/file_watcher.go:1]: I'm referring to below functions:
* `getKeyWatcher`
* `getKey`
specifically, what are the go routines doing?

Comment (dharmit) [reply pkg/watch/file_watcher.go:1]: Thanks. I just read them and it makes a huge difference for me personally. The explanations are very helpful!

**Comment by dharmit** [comment]
/lgtm
/approve

Adding both flags because @rm3l already lgtm'd, but my requests removed the flag.
Thanks @feloy ! 🎉

**Comment by feloy** [comment]
/override ci/prow/v4.10-integration-e2e
Tests pass on IBM Cloud

**Comment by feloy** [comment]
@anandrkskd I've added a PR to fix the problem of failing integration tests on Windows

**Review by rm3l — APPROVED** [review]

**Comment by feloy** [comment]
/ci/prow/v4.10-integration-e2e
Tests pass on IBM cloud

**Comment by feloy** [comment]
/override ci/prow/v4.10-integration-e2e
Tests pass on IBM cloud

**Pull request merged**
